### PR TITLE
[SPARK-19284][SQL]append to partitioned datasource table should without custom partition location

### PR DESCRIPTION
**Pull request by windpiger**
## What changes were proposed in this pull request?

when we append data to a existed partitioned datasource table, the InsertIntoHadoopFsRelationCommand.getCustomPartitionLocations currently
return the same location with Hive default, it should return None.

## How was this patch tested?

### CODE REVIEW

**Comment by windpiger** [comment]
retest this please

**Comment by windpiger** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala [hunk 1]
```diff
@@ -122,7 +122,6 @@ case class CreateDataSourceTableAsSelectCommand(
   override def run(sparkSession: SparkSession): Seq[Row] = {
     assert(table.tableType != CatalogTableType.VIEW)
     assert(table.provider.isDefined)
-    assert(table.schema.isEmpty)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/createDataSourceTables.scala]: we should move this assert to the branch for when table not exist.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala [hunk 2]
```diff
@@ -412,6 +412,10 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
   }
 
   private def createTable(tableIdent: TableIdentifier): Unit = {
+    df.sparkSession.sessionState.executePlan(createTableCommand(tableIdent)).toRdd
+  }
+
+  private[sql] def createTableCommand(tableIdent: TableIdentifier): CreateTable = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala]: why this change?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala [hunk 3]
```diff
@@ -92,6 +96,47 @@ class PartitionedWriteSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("append data an existed partition in a datasource table," +
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala]: `append data to an existing partitioned table without custom partition path`

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala [hunk 4]
```diff
@@ -92,6 +96,47 @@ class PartitionedWriteSuite extends QueryTest with SharedSQLContext {
...
+    "custom location sent to Task should be None ") {
+    withTable("t") {
+      Seq((1, 2)).toDF("a", "b").write.partitionBy("b").saveAsTable("t")
+      val writer = Seq((3, 2)).toDF("a", "b").write.mode("append").partitionBy("b")
+
+      spark.sessionState.executePlan(writer.createTableCommand(TableIdentifier("t")))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala]: how about we make a custom `FileCommitProtocol` to test this behavior?

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala (hunk 3 above)
Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala]: thanks~

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala (hunk 4 above)
Comment (windpiger) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala]: good idea, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala [hunk 5]
```diff
@@ -92,6 +111,16 @@ class PartitionedWriteSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("append data to an existed partitioned table without custom partition path") {
+    withTable("t") {
+      withSQLConf("spark.sql.sources.commitProtocolClass" ->
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/PartitionedWriteSuite.scala]: nit: SQLConf.FILE_COMMIT_PROTOCOL_CLASS.key -> classOf[OnlyDetectCustomPathFileCommitProtocol].getName

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM, pending test

**Comment by cloud-fan** [comment]
thanks, merging to master!
